staging/src/k8s.io/apiserver/pkg/cel/mutation/unstructured/typeresolver_test.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package unstructured
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/google/cel-go/cel"
+
+	"k8s.io/apimachinery/pkg/util/version"
+	"k8s.io/apiserver/pkg/cel/environment"
+	"k8s.io/apiserver/pkg/cel/mutation"
+)
+
+func TestTypeProvider(t *testing.T) {
+	for _, tc := range []struct {
+		name          string
+		expression    string
+		expectedValue any
+	}{
+		{
+			name:          "not an object",
+			expression:    `string(114514)`,
+			expectedValue: "114514",
+		},
+		{
+			name:          "empty",
+			expression:    "Object{}",
+			expectedValue: map[string]any{},
+		},
+		{
+			name:       "Object.spec",
+			expression: "Object{spec: Object.spec{replicas: 3}}",
+			expectedValue: map[string]any{
+				"spec": map[string]any{
+					// an integer maps to int64
+					"replicas": int64(3),
+				},
+			},
```

Comment:
This was exactly the type of test case I was hoping to see. Can we expand this out to include values of different types (scalars, maps, lists, ...) and the use of optionals?

For example `Object{spec: Object.spec{listOfObj: [Object.spec.listOfObj{id: 1}, Object.spec.listOfObj{id: 2}]}}`

Reply:
Good point! I've added test for lists. Scalars and maps should be already covered. Please note that the tests are against the type provider despite having to invoke the CEL complication to cover the accompany EnvOption. We can have more real-world tests in higher level tests once we finish up the whole syntax.

About support of Optional types, these are planned in the second stage. The plan is to filter them out from the returning Apply configuration. I would suggest we have more implementation around `optional` in a separate PR. Would you agree with me? Thank you.

Reply:
Ping sent. See `optional_test.go`